src/main/java/com/offz/spigot/staminaclimb/stamina/StaminaTask.kt

```diff
@@ -6,53 +6,107 @@ import com.offz.spigot.staminaclimb.climbing.ClimbBehaviour
+                    StaminaBar.addProgressWithDeltaTime(StaminaConfig.data.staminaRegenInAir
+                        .coerceAtMost(StaminaConfig.data.barMax), uuid)
+            } else { //Lose stamina
+                if(isPlayerMoving(player)) {
+                    StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemoveWhileMoving * currentWallMultiplier, uuid)
+                }
+                else {
+                    StaminaBar.removeProgressWithDeltaTime(StaminaConfig.data.staminaRemovePerTick * currentWallMultiplier, uuid)
+                }
+            }
 
-            if (progress <= StaminaConfig.data.barRed) { //Changing bar colors and effects on player depending on its progress
+            if (progress <= StaminaConfig.data.barMin) { //Changing bar colors and effects on player depending on its progress
                 bar.color = BarColor.RED
                 bar.setTitle("&c&lStamina".color()) //Make Stamina title red
-                if (uuid.isClimbing) player.stopClimbing()
-
-                uuid.canClimb = false //If player reaches red zone, they can't climb until they get back in green zone
-                player.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 110, 2, false, false))
-                player.addPotionEffect(PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false))
-            } else if (progress < 1 && !uuid.canClimb) {
+                if (uuid.isClimbing) {
+                    player.stopClimbing()
+                    uuid.canClimb = false //If player reaches red zone, they can't climb until they get back in green zone
+                    player.addPotionEffect(PotionEffect(PotionEffectType.SLOW, 110, 2, false, false))
+                    player.addPotionEffect(PotionEffect(PotionEffectType.WEAKNESS, 110, 2, false, false))
+                }
+            } else if (progress < StaminaConfig.data.barMax && !uuid.canClimb) {
                 bar.color = BarColor.RED //Keep Stamina Bar red even in yellow zone while it's regenerating
+            } else if (uuid.isClimbing && progress < StaminaConfig.data.barFlashPoint / 2.0) {
+                ticksSinceLastColorSwitch++
+                if (ticksSinceLastColorSwitch >= ticksToSwitchColor / 2) {
+                    if(bar.color == BarColor.RED)
+                        bar.color = BarColor.YELLOW
+                    else
+                        bar.color = BarColor.RED
+                    ticksSinceLastColorSwitch = 0
+                }
+            } else if (uuid.isClimbing && progress < StaminaConfig.data.barFlashPoint) {
+                ticksSinceLastColorSwitch++
+                if (ticksSinceLastColorSwitch >= ticksToSwitchColor) {
+                    if(bar.color == BarColor.RED)
+                        bar.color = BarColor.GREEN
+                    else
+                        bar.color = BarColor.RED
+                    ticksSinceLastColorSwitch = 0
+                }
             } else {
```

Comment:
This isn't your fault but my brain is absolutely fried reading this file, I might add a commit to this PR to clean it up a bit, but point is I'm not 100% sure of the logic going on here anymore so might have missed something reviewing this